File: Baekjoon/15656.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int n, m;
int arr[10];
int num[10];

void func(int k) {
    // k == m에 도달할 경우 다 출력하고 끝
    if (k == m) {
        for (int i = 0; i < m; i++) {
            cout << num[arr[i]] << ' ';
        }
        cout << '\n';
        return;
    }
    // 아직 채워야 할 공간이 있다면
    for (int i = 0; i < n; i++) {
        arr[k] = i;
        func(k + 1);
    }

    // 0 -> 0 0 -> 0 0 0
    // 0 -> 0 1 -> 0 0
}

int main() {

    cin >> n >> m;
    // 같은 수 여러번 골라도 됨

    for (int i = 0; i < n; i++) {
        cin >> num[i];
    }
    sort(num, num + n);
    func(0);
}

File: Baekjoon/15656.py
# 실버 3 15656. N과 M (7)

# N개의 자연수 중에서 M개를 고른 수열
# 같은 수를 여러 번 골라도 된다

def backtrack(arr, selected, n, m):
    if len(selected) == m:
        for s in selected:
            print(arr[s], end=' ')
        print()
    else:
        for i in range(n):
            selected.append(i)
            backtrack(arr, selected, n, m)
            selected.pop()

n, m = map(int, input().split())

arr = list(map(int, input().split()))
arr.sort()

backtrack(arr, [], n, m)
